### ISSUE TITLE
I wish for a more aggressive "autosubmit" label

### ISSUE THREAD

**Issue by matanlurey** [flutter/flutter#132606]
As written, `autosubmit` is only suitable when the only blocker is the tree is red.

I'd love to be able to use it as a "submit this once I have approval(s) and all checks are green". For example, I might be iterating on a clang tidy issue, or might be waiting for an explicit LGTM - I'd love to not have to manually remember to merge (when the tree is green).

**Comment by ricardoamador** [comment]
It does this already.

**Comment by matanlurey** [comment]
That doesn't match my experience, if a test fails for example the label is removed.

(Whether the test is a legitimate failure, or flake, or infra problem - it doesn't matter)

**Comment by ricardoamador** [comment]
Can you provide an example of this. It should be holding until the status is green and it will then merge the PR.

**Comment by ricardoamador** [comment]
If all checks are green and you are waiting on an explicit LGTM it will hold. It is doing as intended by notifying you of a failure.

**Comment by matanlurey** [comment]
![image](https://github.com/flutter/flutter/assets/168174/939d476f-02f9-450b-a3e8-075da0a65161)

In this case, I wish the label just stuck around until the flake was resolved.

**Comment by ricardoamador** [comment]
You have a test issue there. It means you need to investigate why it failed. This is working as intended.

**Comment by matanlurey** [comment]
But why does it have to remove the label? I'd prefer to just fix the test issue and then have the bot land it.

Otherwise I have to:

1. Investigate and fix (obviously in either case)
2. Re-add the `autosubmit` label
3. ... something else could flake, and I start back at 1

**Comment by ricardoamador** [comment]
For some context:

We remove the label so we do not have to continue processing the Pull Request. So if you add the label and let the pull request sit for days before investigating we would have to continually process the pull request and check to see if you had rerun the check. 

Under the hood we make checks against the github api which costs us quota. We only have so much quota per hour and support repos in flutter as well as dart-lang. Meaning that if we leave the label for every pull request that had the label added at one point, we would quickly burn through that quota and would not be able to process incoming new Pull Requests. Currently we do run up against the quota but that is rare. There are 151 open pull requests in flutter/flutter alone which run 68 checks a piece.

So in order to keep Autosubmit stateless and not hit the quota we remove the label. Hope that helps. This is not to say that we cannot do something different in the future but the current situation is as described.

**Comment by matanlurey** [comment]
I wasn't aware this was cost/load-bearing, so thanks for the explanation.

I guess the root problem is not really autosubmit, it's that I basically _need_ autosubmit or I need to spend manual time doing what autosubmit does for me. So anything that reduces the need for autosubmit _or_ lets me use autosubmit more aggressively would be fine.

**Comment by ricardoamador** [comment]
Yeah no problem. Let me be clear: I am not opposed to this so we can leave this open as we are always looking for ways to improve the functionality but as it stands the main contention is Github quota. Which we will ultimately need to deal with as we expand this tool.

**Comment by matanlurey** [comment]
Btw here is an example that _just_ occurred over at https://github.com/flutter/engine/pull/44730:
![image](https://github.com/flutter/flutter/assets/168174/a2a8cf7a-2c97-43cf-af0a-65356fd01e25)

> xcodebuild: error: Unable to find a destination matching the provided destination specifier:
>		{ name:IosUnitTestsSimulator }
>
>	Ineligible destinations for the "IosUnitTests" scheme:
>		{ platform:iOS, id:dvtdevice-DVTiPhonePlaceholder-iphoneos:placeholder, name:Any iOS Device, error:iOS 16.4 is > not installed. To use with Xcode, first download and install the platform }

It[ appears to be a flake](https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20Engine%20Drone/331151/overview), so I can't really do anything about it but (manually) check that it failed, (manually) check that it was probably a flake, (manually) re-run it, and (manually) re-add the `autosubmit` label.

**Comment by jonahwilliams** [comment]
if we had more aggressive surfacing of (what I would consider) infrastructure failures, then the autosubmit bot could probably more aggressively deflake too.

**Comment by zanderso** [comment]
We likely need an infra ticket queue issue filed for the flake mentioned in https://github.com/flutter/flutter/issues/132606#issuecomment-1679779452. That looks like a rogue bot that doesn't have the right iOS simulator(s) installed, maybe?

This issue is similar to the request for deflaking presubs on autoroller PRs here: https://github.com/flutter/flutter/issues/126556. I'm not sure if this is a good idea, but maybe for the non-autoroller case, we could have a human-added 'deflake' label that would change the behavior of the autosubmit bot (re-run failed presubs, keep the 'autosubmit' label).

**Comment by matanlurey** [comment]
> We likely need an infra ticket queue issue filed for the flake mentioned in [#132606 (comment)](https://github.com/flutter/flutter/issues/132606#issuecomment-1679779452).

Yup, re-opened https://github.com/flutter/flutter/issues/125273#issuecomment-1679781221.

**Comment by godofredoc** [comment]
https://github.com/flutter/engine/pull/44759#issuecomment-1680860979 failed because clang tidy ran for over an hour and it didn't complete, these kind of issues are difficult to categorize as infra or PR issues without manual inspection.

So far the comments in this bug can be categorized as flakes in the build/test process (not infra) and properly propagate subbuild timeouts to main build (infra tracking bug https://github.com/flutter/flutter/issues/123645)

**Comment by jonahwilliams** [comment]
While I agree that clang tidy timing out is not really an infra issue, its also almost certainly not related to the PR in question.

@godofredoc I'm not sure why Xcode not having an installed simualtor runtime wouldn't be an infra issue?

**Comment by matanlurey** [comment]
> these kind of issues are difficult to categorize as infra or PR issues without manual inspection

And that was kind of my original point - I'd like to use robots to retry instead of manual inspection :)

**Comment by godofredoc** [comment]
@jonahwilliams I believe @keyonghan has already a fix for the xcode runtimes.

**Comment by keyonghan** [comment]
> @jonahwilliams I believe @keyonghan has already a fix for the xcode runtimes.

Issue is tracked https://github.com/flutter/flutter/issues/125273, the fix (https://flutter-review.googlesource.com/c/recipes/+/49482) is under review, and the problem bot has been refreshed: https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/keyonghan_google.com/982451e4ac666d4e859ba6c30be72d9de67c4d245db6a6d9e3d816f0be58735a/+/build.proto

**Comment by CaseyHillers** [comment]
Have we looked into [Conditional Requests](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#conditional-requests)? That would prevent any API usage if nothing has changed, but still allow us to keep the autosubmit label attached.

**Comment by ricardoamador** [comment]
That could be really useful across the board in our services.

**Comment by matanlurey** [comment]
I would love for us to evaluate if we could use auto-merge instead of trying to replicate that feature set in auto-submit, mostly because auto-merge is (a) built-in to GitHub, and (b) requires no additional resources/quota.

**Comment by ricardoamador** [comment]
You referring to this comment: 
![image](https://github.com/flutter/flutter/assets/32242716/aa84a618-4af7-4ed6-b748-c3434db72f6c)

I am sure this works well on your personal projects. It might work here however we use Autosubmit for more than simply validating check runs and waiting to merge. @CaseyHillers is correct that auto-merge defers to branch protection rules which is not as strict as what auto-submit is enforcing.

Quota can be an issue if you are simply keeping the autosubmit label on the pull request. We remove the label so developers can also take action on issues that arise with the pull request and prevent busy waiting.

**Comment by matanlurey** [comment]
Could we move our custom branch protection rules to a _check_ (i.e. turns the PR red) instead of making it private to the bot?

If we did that, we wouldn't need auto-submit bot at all (AFAICT), and would not have the same quota issues we have today?

**Comment by matanlurey** [comment]
(As I mostly joked in Discord, the alternative is I write my own bot that automatically re-adds the label, and that doesn't seem right to me)
